### ISSUE TITLE
Add option to disable automatic barriers

### ISSUE THREAD

**Issue by stgeke** [libocca/occa#484]

**Comment by kris-rowe** [comment]
To clarify the intended usage: are you looking for a way to disable all automatic barriers for a kernel&mdash;e.g., by passing kernel properties, or specific automatic barriers within a kernel&mdash;e.g., using an attribute like `@nobarrier` similar to the `nowait` directive in OpenMP?

**Comment by stgeke** [comment]
Currently a `@barrier` is added after an `inner` block. It would be nice to have an option to disable this.

**Comment by noelchalmers** [comment]
A small clarification is that a barrier is added automatically after an inner block only if that block used shmem at all, and it's not the last inner block. 

We did used to have the option to disable auto barrier addition, but I agree with Kris that this probably isn't ideal since it's a pretty heavy toggle to set in entire okl file(s). These days I just fuse inner blocks when I dont want the barrier.

**Comment by stgeke** [comment]
A related issue: If the inner size <= warpSize a warp-wide barrier should be added. Currently no `@barrier` is added at all. That's tricky at least for Nvidia's Volta and later architectures (you can no longer assume that the threads in a wrap run in lock-step).

**Comment by noelchalmers** [comment]
Can you give an example of this? Do you have individual lanes of the warp trying to communicate through global memory? I haven't seen any use for __syncwarp aside from that scenario. 

The normal __syncthreads is identical to __syncwarp when inner size <= warp size.

**Comment by stgeke** [comment]
My bigger concern is that at the moment no barriers are added at all. Maybe I recall incorrectly?
Isn't  __syncwarp faster than __syncthreads?

**Comment by noelchalmers** [comment]
There's no syncwarps added currently, that's correct. But a barrier like that should only be added when such a barrier is needed. I'm curious where specifically you think the barrier is needed. Right now you could obviously just rely on splitting inner blocks and getting coherency through the usual syncthreads.

Is syncwarp faster than syncthreads? Depends on the usage. They're likely comparable in time if you have to wait on global mem fences. If the threadblock is truly made of warps that dont share data with one another (so syncthreads isnt needed), but _do_ share data between the lanes of the warp, then yes there's probably opportunity to progress some warps while barriering others. I dont think that's common, however. Is that what you need to happen?

**Comment by kris-rowe** [comment]
> A related issue: If the inner size <= warpSize a warp-wide barrier should be added. Currently no `@barrier` is added at all. That's tricky at least for Nvidia's Volta and later architectures (you can no longer assume that the threads in a wrap run in lock-step).

This is also relevant for OpenCL and SYCL/DPC++ since the innermost `@inner` loop will be mapped to a sub-group. The new versions of the standards support sub-group barriers. I have opened a separate issue (#516 ) for this.
